### PR TITLE
[graalvm] Make compatible with graalvm native image

### PR DESCRIPTION
**Pull request by ericdallo**
Before submitting a PR make sure the following things have been done:

- [X] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [ ] You've added tests to cover your change(s)
- [X] All tests are passing
- [X] The new code is not generating reflection warnings
- [ ] You've updated the [changelog](../CHANGELOG.md)(that only applies to user-visible changes)

I was hacking to use nrepl on clojure-lsp as an enhancement, and for that would be necessary to nrepl be graalvm compile compatible, I realized that to achieve that would need only a few changes, so this PR addresses that:
- Change some top-level `def`s to `defn` as graal tries to analyze that code at build time instead of runtime as this is a recommendation for Clojure graalvm compiled libraries.

I tested a simple `nrepl.core/connect` and sent a simple eval and everything worked, maybe we may need to tweak a little more later for full nrepl compatibility

### CODE REVIEW

File: src/clojure/nrepl/middleware/session.clj [hunk 1]
```diff
@@ -85,7 +85,10 @@
                          ^BlockingQueue queue
                          thread-factory)))
 
-(def default-executor "Delay containing the default Executor." (delay (configure-executor)))
+(defn default-executor
+  "Delay containing the default Executor."
+  []
+  (delay (configure-executor)))
```
Comment (bbatsov) [inline src/clojure/nrepl/middleware/session.clj:91]: I think this `delay` is probably no longer needed, as it won't be evaluated right away anyways.

**Comment by bbatsov** [comment]
> Change some top-level defs to defn as graal tries to analyze that code at build time instead of runtime as this is a recommendation for Clojure graalvm compiled libraries.

I guess you should add this to your commit message, so it's clearer why those things needed to be changed. Please, add a changelog entry about this as well. 

The first change is technically speaking a breaking change, but I looked in `cider-nrepl` and `refactor-nrepl` and I don't see any usages. In general it's unlikely that anyone's using this directly, so the change should be safe.

**Comment by borkdude** [comment]
@ericdallo What was the error you got without changes?

You can leave the delay as a delay since a delay is basically a function that caches itself on first invocation.
The same holds for other `def` -> `defn` changes: you can turn these into delays instead of functions and it should work, provided the delays are not initialized at build time. So you should turn each reference of `(def foo bar)` into `(def foo (delay @bar))`.

**Comment by ericdallo** [comment]
Good to know, I was having issues with Thread being initialized at build time, I'll change to delay and try again

**Comment by ericdallo** [comment]
@bbatsov I removed the breaking changes following the @borkdude suggestions and it worked well with clojure-lsp, I created this  sample project to test nrepl compilation with graalvm: https://github.com/ericdallo/nrepl-graalvm-sample but is not working for some reason 😅 
<details>
<summary>A new exception happens</summary>
<pre>
Error: #error {
 :cause [B
 :via
 [{:type java.lang.ClassNotFoundException
   :message [B
   :at [java.lang.Class forName DynamicHub.java 1433]}]
 :trace
 [[java.lang.Class forName DynamicHub.java 1433]
  [clojure.lang.RT classForName RT.java 2212]
  [clojure.lang.RT classForName RT.java 2221]
  [nrepl.bencode$fn__206$fn__207 invoke bencode.clj 316]
  [clojure.lang.MultiFn invoke MultiFn.java 233]
  [nrepl.transport$safe_write_bencode invokeStatic transport.clj 111]
  [nrepl.transport$safe_write_bencode invoke transport.clj 103]
  [nrepl.transport$bencode$fn__528 invoke transport.clj 130]
  [nrepl.transport.FnTransport send transport.clj 34]
  [nrepl.core$client$this__622 invoke core.clj 53]
  [clojure.core$comp$fn__5825 invoke core.clj 2573]
  [clojure.core$comp$fn__5825 invoke core.clj 2573]
  [clojure.lang.AFn applyToHelper AFn.java 154]
  [clojure.lang.RestFn applyTo RestFn.java 132]
  [clojure.lang.AFunction$1 doInvoke AFunction.java 31]
  [clojure.lang.RestFn invoke RestFn.java 408]
  [nrepl.core$message invokeStatic core.clj 93]
  [nrepl.core$message invoke core.clj 85]
  [ericdallo.nrepl_graalvm_sample$test_eval invokeStatic nrepl_graalvm_sample.clj 17]
  [ericdallo.nrepl_graalvm_sample$test_eval invoke nrepl_graalvm_sample.clj 12]
  [ericdallo.nrepl_graalvm_sample$_main$fn__689 invoke nrepl_graalvm_sample.clj 22]
  [ericdallo.nrepl_graalvm_sample$_main invokeStatic nrepl_graalvm_sample.clj 21]
  [ericdallo.nrepl_graalvm_sample$_main doInvoke nrepl_graalvm_sample.clj 20]
  [clojure.lang.RestFn invoke RestFn.java 397]
  [clojure.lang.AFn applyToHelper AFn.java 152]
  [clojure.lang.RestFn applyTo RestFn.java 132]
  [ericdallo.nrepl_graalvm_sample main nil -1]]}

</pre>
</details>

**Comment by borkdude** [comment]
Followed by this commit: https://github.com/nrepl/bencode/commit/96ea1854b65359926b737a4decb87a683b2532be

**Comment by ericdallo** [comment]
Thanks, I found that line but I had no clue how to fix it, I'll do those changes so!

**Comment by ericdallo** [comment]
Done, everything working good on the sample project as well, thank you for the help @borkdude

File: src/clojure/nrepl/util/completion.clj [hunk 2]
```diff
@@ -168,16 +168,18 @@
   (.. file (replace ".class" "") (replace File/separator ".")))
 
 (def top-level-classes
-  (misc/noisy-future
-   (doall
-    (for [file classfiles :when (re-find #"^[^\$]+\.class" file)]
-      (classname file)))))
+  (delay
```
Comment (bbatsov) [inline src/clojure/nrepl/util/completion.clj:171]: Just add some note about the `delay`s, otherwise someone might optimize them away. :-)

Comment (ericdallo) [reply src/clojure/nrepl/util/completion.clj:172]: Done @bbatsov

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**